packages/cli/src/commands/hydrogen/upgrade.ts

```diff
@@ -0,0 +1,2931 @@
+
+  if (typeof rawPackageJson !== 'string') {
+    renderFatalError({
+      name: 'error',
+      type: 0,
+      message: 'Could not find a valid package.json',
+      tryMessage: `Please make sure you are running the command in a npm project`,
+    });
+
+    if (!process.env.SHOPIFY_UNIT_TEST) {
+      process.exit(0);
+    }
+
+    return;
+  }
+
+  const packageJson = JSON.parse(rawPackageJson) as PackageJson;
+
+  const currentDependencies = {
+    ...(packageJson?.dependencies ?? {}),
+    ...(packageJson?.devDependencies ?? {}),
+  } as Dependencies;
+
+  const currentVersion = currentDependencies?.['@shopify/hydrogen'];
+
+  if (!currentVersion) {
+    renderFatalError({
+      name: 'error',
+      type: 0,
+      message: 'Could not find a valid Hydrogen version in package.json',
+      tryMessage: `Please make sure you are running the command in a Hydrogen project`,
+    });
+
+    if (!process.env.SHOPIFY_UNIT_TEST) {
+      process.exit(0);
+    }
+  }
+
+  return {currentVersion, currentDependencies};
+}
+
+/**
+ * (Temp) Mock-fetches the changelog-versions.json while we merge
+ */
+export async function fetchTempChangelog(): Promise<ChangeLog> {
+  return Promise.resolve(TEMP_CHANGELOG);
+}
+
+/**
+ * Fetches the changelog.json file from the Hydrogen repo
+ */
+export async function fetchChangelog(): Promise<ChangeLog | undefined> {
+  if (CACHED_CHANGELOG) {
+    return CACHED_CHANGELOG;
+  }
+  try {
+    const response = await fetch(
+      // TODO: https://github.com/Shopify/hydrogen-shopify-dev/pull/154
+      // TODO: https://github.com/Shopify/hydrogen/blob/main/packages/cli/src/changelog.json
+      'https://hydrogen.shopify.dev/changelog.json',
+    );
+
+    if (!response.ok) {
+      throw new Error('Failed to fetch changelog.json');
+    }
+
+    const json = (await response.json()) as object;
+
+    if ('releases' in json && 'url' in json) {
+      CACHED_CHANGELOG = json as ChangeLog;
+      return CACHED_CHANGELOG;
+    }
+  } catch (error) {
+    renderFatalError({
+      name: 'error',
+      type: 0,
+      message: 'Failed to fetch changelog',
+      tryMessage: `Please try again later`,
+    });
+
+    if (!process.env.SHOPIFY_UNIT_TEST) {
+      process.exit(0);
+    }
+  }
+}
+
+/**
+ * Gets the list of available upgrades based on the current version
+ */
+export function getAvailableUpgrades({
+  releases,
+  currentVersion,
+}: {
+  releases: ChangeLog['releases'];
+  currentVersion: string;
+}) {
+  const currentPinnedVersion = getAbsoluteVersion(currentVersion);
+  let currentMajorVersion = '';
+
+  const availableUpgrades = releases.filter((release) => {
+    const isUpgrade = semver.gt(release.version, currentPinnedVersion);
+    if (!isUpgrade) return false;
+    if (currentMajorVersion !== release.version) {
+      currentMajorVersion = release.version;
+      return isUpgrade;
+    }
+    return false;
+  }) as Array<Release>;
```

Comment:
Maybe I'm missing something but as I read this, I think we need more checks to know if a release is an upgrade for the current version.

Say you have hydrogen@X, and X is its latest version. However, we have several releases of CLI or remix-oxygen for hydrogen@X, and you don't have the latest versions installed for those packages.
This would tell you that you are up to date with hydrogen@X and it won't update CLI or remix-oxygen, right?

I think to be sure something is or is not an upgrade we'd need to compare every package version in the release to the local versions :thinking: